fault_proof/bin/proposer.rs

```diff
@@ -0,0 +1,327 @@
+        let oldest_game_index = latest_game_index
+            .saturating_sub(U256::from(self.config.max_games_to_check_for_resolution));
+        let games_to_check =
+            latest_game_index.min(U256::from(self.config.max_games_to_check_for_resolution));
+
+        let (should_attempt_resolution, game_address) =
+            self.should_attempt_resolution(oldest_game_index).await?;
+
+        if should_attempt_resolution {
+            for i in 0..games_to_check.to::<u64>() {
+                let index = oldest_game_index + U256::from(i);
+                self.try_resolve_unchallenged_game(index).await?;
+            }
+        } else {
+            tracing::info!(
+                "Oldest game {:?} at index {:?} is not resolved, not attempting resolution",
+                game_address,
+                oldest_game_index
+            );
+        }
+
+        Ok(())
+    }
+
+    /// Runs the proposer indefinitely.
+    async fn run(&self) -> Result<()> {
+        tracing::info!("OP Succinct Proposer running...");
+        let mut interval = time::interval(Duration::from_secs(self.config.fetch_interval));
+
+        loop {
+            interval.tick().await;
+
+            let _span = tracing::info_span!("[[Proposing]]").entered();
+
+            let safe_l2_head_block_number = self
+                .l2_provider
+                .get_l2_block_by_number(BlockNumberOrTag::Safe)
+                .await?
+                .header
+                .number;
+            tracing::debug!("Safe L2 head block number: {:?}", safe_l2_head_block_number);
+
+            let latest_valid_proposal = self
+                .factory
+                .get_latest_valid_proposal(self.l1_provider.clone(), self.l2_provider.clone())
+                .await?;
+
+            let (next_l2_block_number_for_proposal, parent_game_index) = match latest_valid_proposal
+            {
+                Some((latest_block, latest_game_idx)) => (
+                    latest_block + U256::from(self.config.proposal_interval_in_blocks),
+                    latest_game_idx.to::<u32>(),
+                ),
+                None => {
+                    // For first game, start from genesis L2 block number + proposal interval
+                    let genesis_l2_block_number = self
+                        .factory
+                        .get_genesis_l2_block_number(
+                            self.config.game_type,
+                            self.l1_provider.clone(),
+                        )
+                        .await?;
+
+                    (
+                        genesis_l2_block_number
+                            .checked_add(U256::from(self.config.proposal_interval_in_blocks))
+                            .unwrap(),
+                        u32::MAX, // Use max value for first game's parent index
+                    )
+                }
+            };
+
+            if U256::from(safe_l2_head_block_number) > next_l2_block_number_for_proposal {
```

Comment:
Here you check if the `next_l2_block_number_for_proposal` can be proposed. If it is greater than the safe_head, you cannot propose.